amd_openvx_extensions/amd_nn/src/cast_layer.cpp

```diff
@@ -0,0 +1,241 @@
+}
+
+static vx_status VX_CALLBACK opencl_codegen(
+    vx_node node,                                  // [input] node
+    const vx_reference parameters[],               // [input] parameters
+    vx_uint32 num,                                 // [input] number of parameters
+    bool opencl_load_function,                     // [input]  false: normal OpenCL kernel; true: reserved
+    char opencl_kernel_function_name[64],          // [output] kernel_name for clCreateKernel()
+    std::string& opencl_kernel_code,               // [output] string for clCreateProgramWithSource()
+    std::string& opencl_build_options,             // [output] options for clBuildProgram()
+    vx_uint32& opencl_work_dim,                    // [output] work_dim for clEnqueueNDRangeKernel()
+    vx_size opencl_global_work[],                  // [output] global_work[] for clEnqueueNDRangeKernel()
+    vx_size opencl_local_work[],                   // [output] local_work[] for clEnqueueNDRangeKernel()
+    vx_uint32& opencl_local_buffer_usage_mask,     // [output] reserved: must be ZERO
+    vx_uint32& opencl_local_buffer_size_in_bytes   // [output] reserved: must be ZERO
+)
+{
+	vx_enum input_type, output_type;
+    vx_size num_dims;
+    vx_size input_dims[4],  output_dims[4];
+
+    ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_NUMBER_OF_DIMS, &num_dims, sizeof(num_dims)));
+    ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_DATA_TYPE, &input_type, sizeof(input_type)));
+    ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_DIMS, input_dims, sizeof(input_dims)));
+
+    vx_int32 to;
+    ERROR_CHECK_STATUS(vxCopyScalar((vx_scalar)parameters[1], &to, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+
+    ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_NUMBER_OF_DIMS, &num_dims, sizeof(num_dims)));
+    ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_DATA_TYPE, &output_type, sizeof(output_type)));
+    ERROR_CHECK_STATUS(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_DIMS, output_dims, sizeof(output_dims)));
+
+#if ENABLE_DEBUG_PRINT_DIMS
+    if (num_dims == 4) { 
+    	std::cout << "cast_layer input " << input_dims[3] << " " << input_dims[2] << " " << input_dims[1] << " " << input_dims[0] << " ";
+    	std::cout << "cast_layer output " << output_dims[3] << " " << output_dims[2] << " " << output_dims[1] << " " << output_dims[0] << std::endl;
+    }
+    else if(num_dims == 2) {
+    	std::cout << "cast_layer input " << " " << input_dims[1] << " " << input_dims[0] << " ";
+    	std::cout << "cast_layer output " << " " << output_dims[1] << " " << output_dims[0] << std::endl;
+    }
+    
+#endif
+
+    vx_size tot_count_elements = input_dims[0]*input_dims[1];
+    bool input_element_count_multiple_of_4 = (tot_count_elements & 3) ? false : true;
+
+	strcpy(opencl_kernel_function_name, "cast_layer");
+
+    opencl_work_dim = 2;
+    opencl_global_work[0] = input_dims[0];
```

Comment:
How about input_dims[2] and input_dims[3] in case of 4D tensors? Looks like global_work doesn't process those elements